test/widgets/subscription_list_test.dart

```diff
@@ -89,18 +89,71 @@ void main() {
     check(isUnpinnedHeaderInTree()).isTrue();
   });
 
-  testWidgets('subscription sort', (tester) async {
+  testWidgets('pinned subscriptions are sorted', (tester) async {
     await setupStreamListPage(tester, subscriptions: [
       eg.subscription(eg.stream(streamId: 1, name: 'd'), pinToTop: true),
-      eg.subscription(eg.stream(streamId: 2, name: 'c'), pinToTop: false),
+      eg.subscription(eg.stream(streamId: 2, name: 'c'), pinToTop: true),
       eg.subscription(eg.stream(streamId: 3, name: 'b'), pinToTop: true),
+      eg.subscription(eg.stream(streamId: 4, name: 'a'), pinToTop: true),
+    ]);
+    check(isPinnedHeaderInTree()).isTrue();
+
+    final streamListItems = tester.widgetList<SubscriptionItem>(find.byType(SubscriptionItem)).toList();
+    check(streamListItems.map((e) => e.subscription.streamId)).deepEquals([4, 3, 2, 1]);
+  });
+
+  testWidgets('pinned subscriptions are sorted case insensitively', (tester) async {
+    await setupStreamListPage(tester, subscriptions: [
+      eg.subscription(eg.stream(streamId: 1, name: 'D'), pinToTop: true),
+      eg.subscription(eg.stream(streamId: 2, name: 'c'), pinToTop: true),
+      eg.subscription(eg.stream(streamId: 3, name: 'B'), pinToTop: true),
+      eg.subscription(eg.stream(streamId: 4, name: 'a'), pinToTop: true),
+    ]);
+    check(isPinnedHeaderInTree()).isTrue();
+
+    final streamListItems = tester.widgetList<SubscriptionItem>(find.byType(SubscriptionItem)).toList();
+    check(streamListItems.map((e) => e.subscription.streamId)).deepEquals([4, 3, 2, 1]);
+  });
+
+  testWidgets('unpinned subscriptions are sorted', (tester) async {
+    await setupStreamListPage(tester, subscriptions: [
+      eg.subscription(eg.stream(streamId: 1, name: 'd'), pinToTop: false),
+      eg.subscription(eg.stream(streamId: 2, name: 'c'), pinToTop: false),
+      eg.subscription(eg.stream(streamId: 3, name: 'b'), pinToTop: false),
       eg.subscription(eg.stream(streamId: 4, name: 'a'), pinToTop: false),
     ]);
+    check(isUnpinnedHeaderInTree()).isTrue();
+
+    final streamListItems = tester.widgetList<SubscriptionItem>(find.byType(SubscriptionItem)).toList();
+    check(streamListItems.map((e) => e.subscription.streamId)).deepEquals([4, 3, 2, 1]);
+  });
+
+  testWidgets('unpinned subscriptions are sorted case insensitively', (tester) async {
+    await setupStreamListPage(tester, subscriptions: [
+      eg.subscription(eg.stream(streamId: 1, name: 'D'), pinToTop: false),
+      eg.subscription(eg.stream(streamId: 2, name: 'c'), pinToTop: false),
+      eg.subscription(eg.stream(streamId: 3, name: 'B'), pinToTop: false),
+      eg.subscription(eg.stream(streamId: 4, name: 'a'), pinToTop: false),
+    ]);
+    check(isUnpinnedHeaderInTree()).isTrue();
+
+    final streamListItems = tester.widgetList<SubscriptionItem>(find.byType(SubscriptionItem)).toList();
+    check(streamListItems.map((e) => e.subscription.streamId)).deepEquals([4, 3, 2, 1]);
```

Comment:
Another part of keeping each test case focused and concise would be to factor out the definition of `streamListItems`, plus the map to stream IDs, as a little helper function. Then each of these test cases can just say something like:
```dart
    check(listedStreamIds(tester)).deepEquals([4, 3, 2, 1]);
```

Reply:
I feel like the hardest thing about test code refactoring is to consider when and where to refactor. but I am improving already, Thank you.

Reply:
I feel like the hardest thing about test code refactoring is to consider when and where to refactor. but I am improving already, Thank you.

Reply:
I feel like the hardest thing about test code refactoring is to consider when and where to refactor. but I am improving already, Thank you.